ipyparallel/cluster/shellcmd.py

```diff
@@ -0,0 +1,860 @@
+#!/usr/bin/env python
+"""
+Shell helper application for OS independent shell commands
+
+Currently, the following command are supported:
+* start:   starts a process into background and returns its pid
+* running: check if a process with a given pid is running
+* kill:    kill a process with a given pid
+* mkdir:   creates directory recursively (no error if directory already exists)
+* rmdir:   remove directory recursively
+* exists:  checks if a file/directory exists
+* remove:  removes a file
+"""
+import base64
+import enum
+import inspect
+import json
+import os
+import re
+import shlex
+import sys
+import time
+from datetime import datetime
+from random import randint
+from subprocess import CalledProcessError, Popen, TimeoutExpired, check_output
+from tempfile import NamedTemporaryFile
+
+
+class Platform(enum.Enum):
+    Unknown = (0,)
+    Linux = (1,)
+    Windows = (2,)
+    MacOS = 3
+
+    @staticmethod
+    def get():
+        import sys
+
+        tmp = sys.platform.lower()
+        if tmp == "win32":
+            return Platform.Windows
+        elif tmp == "linux":
+            return Platform.Linux
+        elif tmp == "darwin":
+            return Platform.MacOS
+        else:
+            raise Exception(f"Unknown platform label '{sys.platform}'")
+
+
+class SimpleLog:
+    def __init__(self, filename):
+        userdir = ""
+        platform = Platform.get()
+        if platform == Platform.Windows:
+            userdir = os.environ["USERPROFILE"]
+        elif platform == Platform.Linux:
+            userdir = os.environ["HOME"] if "HOME" in os.environ else ""
+        self.filename = filename.replace(
+            "${userdir}", userdir
+        )  # replace possible ${userdir} placeholder
+
+    def _output_msg(self, level, msg):
+        with open(self.filename, "a") as f:
```

Comment:
Why not hold the file open rather than opening it for every message?

Reply:
I added the SimpleLog class only for debugging reasons, mainly to track down the windows github runner issues (CREATE_BREAKAWAY_FROM_JOB flag not supported in github windows runners). We could basically remove the class and code that uses it, but in case of issues I would need to re-add similare code. The current logger uses a 'stateless' concept without the need of explicit flushing. This way, it is less likely that parallel processes writing to the same file produce interfering messages (at least I think so).  But If you find your suggestion important I will change the code (even though is not used in general)